compiler/resolution/visibleFunctions.cpp

```diff
@@ -449,131 +469,161 @@ static void lookAtTypeFirst(const char* name, CallExpr* call,
+        // When private methods and fields are supported, we'll need to extend
+        // this
+        visibleFns.add(fn);
+      }
+    }
+  }
+}
+
+static bool needToTraverseUse(bool firstVisit, bool inUseChain, bool isPrivate);
+
+static void getVisibleMethodsFromUseList(const char* name, CallExpr* call,
+                                         BlockStmt* block,
+                                         VisibilityInfo* visInfo,
+                                         std::set<BlockStmt*>& visited,
+                                         Vec<FnSymbol*>& visibleFns,
+                                         bool inUseChain, bool firstVisit) {
+  // the block uses other modules
+  for_actuals(expr, block->useList) {
+    SymExpr* se = NULL;
+    if (UseStmt* use = toUseStmt(expr)) {
+      if (!needToTraverseUse(firstVisit, inUseChain, use->isPrivate))
+        continue;
+      if (use->skipSymbolSearch(name))
+        continue;
+      se = toSymExpr(use->src);
+    } else if (ImportStmt* import = toImportStmt(expr)) {
+      if (!needToTraverseUse(firstVisit, inUseChain, import->isPrivate))
+        continue;
+      if (import->skipSymbolSearch(name))
+        continue;
+
+      se = toSymExpr(import->src);
+
+    } else {
+      INT_FATAL("bad expr in useList, expected ImportStmt or UseStmt");
+    }
+    // A use statement could be of an enum instead of a module, but only
+    // modules can define functions.
+    if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
+      if (mod->isVisible(call)) {
+        getVisibleMethodsImpl(name, call, mod->block, visInfo, visited,
+                              visibleFns, true);
+      }
+    }
+  }
 }
 
+
+
 static void getVisibleMethodsImpl(const char* name, CallExpr* call,
-                                  BlockStmt*            block,
-                                  VisibilityInfo*       visInfo,
+                                  BlockStmt* block, VisibilityInfo* visInfo,
                                   std::set<BlockStmt*>& visited,
-                                  Vec<FnSymbol*>&       visibleFns) {
+                                  Vec<FnSymbol*>& visibleFns, bool inUseChain) {
   if (block == rootBlock) return; // nothing there
   //
   // avoid infinite recursion due to modules with mutual uses
   //
-  if (visited.find(block) == visited.end()) {
-
-    BlockStmt* instantiationPt = getVisibleFnsInstantiationPt(block);
-
-    if (call->id == breakOnResolveID) {
-      getVisibleFnsShowBlock("methods", block, instantiationPt);
-    }
+  const bool firstVisit = (visited.find(block) == visited.end());
+  if (!firstVisit && inUseChain) {
+    // We've seen this block already, but we just found it again.  If we found
+    // it again from another use chain, just return.  If we found it from going
+    // up in scope from the call site, that means that we may have skipped
+    // private uses, so we should go through only the private uses - not in a
+    // use chain.
```

Comment:
I found the comment hard to follow. Could it be:
```
    // We've seen this block already, but we just found it again from a use chain.
    //   * If we found it earlier from another use chain, no further action is needed, so return
   //    * If we found it earlier from going up in scope from the call site, no further action is needed, so return
```

(You could say more about private uses in there but I found it confusing - what is a private use not in a use chain?).

Reply:
The boolean doesn't track how we found the scope previously, it tracks how we are currently finding the scope.  There's three scenarios:
- found the scope the first time by traversing a use
  - found the scope again by also traversing a use
  - found the scope again by going up in scope from the call site (this is the only case where we care that we found the scope again)
- found the scope the first time by going up in scope from the call site, and again by traversing a use
(it is not possible to find the scope twice when both times were by going up scope from the call site - is what I was going to type, but the lookAtTypeFirst stuff changes that, now that I think of it . . .)

We don't traverse private use statements unless we found the scope with the use statements by going up in scope from the call site.  And we don't traverse scopes we've already seen before unless there's a reason for us to believe we'll find different symbols this time around.

Reply:
Seems like most of what you just posted would make a great and more understandable comment there :)